qmi/core/transport.py

```diff
@@ -612,155 +635,314 @@ def _safe_socket(self) -> socket.socket:
+        self._safe_socket.settimeout(0)
+        while True:
+            try:
+                b, addr = self._safe_socket.recvfrom(4096)
+            except (BlockingIOError, socket.timeout):
+                # no more bytes available
+                break
+            except OSError:
+                # UDP protocol was used and > 4096 bytes in socket buffer. This discards also the rest of the packet.
+                break
+
+            if not b:
+                # end of stream
+                break
+
+
+class QMI_UdpTransport(QMI_UdpTcpTransportBase):
+    """
+    An instance of QMI_UdpTransport represents a server-side UDP connection with
+    a listening port to an instrument. Client-side UDP connections are not supported.
+
+    The maximum UDP packet size is 64KB (including headers). For `write` and `read` functions,
+    this means limitations. The write function we can check the data size to send and split into
+    multiple messages. For `read` functions this is not possible and the client must limit the
+    data size as necessary. If for `read` functions the receivable data size is larger than
+    the size defined in `recvfrom(<size>)`, an exception is raised.
+    """
+
+    def __init__(self, host: str, port: int) -> None:
+        """Initialize the UDP transport by validating the UDP host and port.
+
+        Parameters:
+            host: The server IP address.
+            port: The port for the address.
+        """
+        self._validate_udp_port(port)
+        super().__init__(host, port)
+
+    @staticmethod
+    def _validate_udp_port(port):
+        if port == QMI_Context.DEFAULT_UDP_RESPONDER_PORT:
+            raise QMI_TransportDescriptorException(f"UDP port number {port} not allowed")
+
+    def __str__(self) -> str:
+        remote_addr = format_address_and_port(self._address)
+        return f"QMI_UdpTransport(remote={remote_addr})"
+
+    def _open_transport(self) -> None:
+        super()._open_transport()
+        # Validate the host exists before trying to create a socket
+        socket.gethostbyname(self._address[0])
+        # Create socket
+        self._socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        # To set the local address to point to our client, we need to bind it.
+        self._socket.bind(("", self._address[1]))
 
     def close(self) -> None:
-        _logger.debug("Closing TCP transport %s", self)
+        _logger.debug("Closing UDP transport %s", self)
         super().close()
-        self._safe_socket.close()
 
     def write(self, data: bytes) -> None:
         self._check_is_open()
         # NOTE: We explicitly adjust the socket timeout before each send/recv call.
         self._safe_socket.settimeout(None)
-        self._safe_socket.sendall(data)
+        self._safe_socket.sendto(data, self._address)
 
-    def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
+    def read(self, nbytes: int, timeout: Optional[float] = None) -> bytes:
```

Comment:
Almost. There's on L776 (UDP)
`                b, addr = self._read_from_socket(max(nbytes - nbuf, 4096))`
and on L892 (TCP)
`                b, addr = self._read_from_socket(nbytes - nbuf)`. 
The unhandy location within the `while` loop prevented me unifying the methods more. Perhaps it can be done, I just didn't see how to do it without changing the method definition. Adding a new input "packet_size: int = 4096" for example could solve this and it could be then moved into the base class. But I think this is scope creep and should be done in a separate ticket.

Reply:
If you change the line to something like this, you should be able to push it to the base class.:
```python
    b, addr = self._read_from_socket(max(nbytes - nbuf, MIN_READ_PACKET_LEN))

class UdpTransport(SocketTransport):
    MIN_READ_PACKET_LEN = 4096

class TcpTransport(SocketTransport):
    MIN_READ_PACKET_LEN = 0
```

Reply:
OK, so for this `read` we would introduce `MIN_PACKET_SIZE` and for `read_until` (comment below) `MAX_PACKET_SIZE`? Yes, then I guess it could be done. In the base class these would all be set to 0 to make the definition of the class variables so that MyPy won't start heckling.